bip78/src/receiver/mod.rs

```diff
@@ -42,23 +53,91 @@ impl UncheckedProposal {
         })
     }
 
+    /// The Sender's Original PSBT
     pub fn get_transaction_to_check_broadcast(&self) -> bitcoin::Transaction {
         self.psbt.clone().extract_tx()
     }
 
-    pub fn assume_broadcastability_was_verified(self) -> UnlockedProposal {
-        UnlockedProposal {
+    /// Receiver MUST check that the Original PSBT from the sender
+    /// can be broadcast, i.e. `testmempoolaccept` bitcoind rpc returns { "allowed": true,.. }
+    /// for `get_transaction_to_check_broadcast()` before calling this method.
+    ///
+    /// Check this if you generate bitcoin uri to receive PayJoin on sender request without manual human approval, like a payment processor.
+    /// Such so called "interactive" receivers are otherwise vulnerable to probing attacks.
+    /// If a sender can make requests at will, they can learn which bitcoin the receiver owns at no cost.
+    /// Broadcasting the Original PSBT after some time in the failure case makes incurs sender cost and prevents probing.
+    ///
+    /// Call this after checking downstream.
+    pub fn attest_tested_and_scheduled_broadcast(self) -> MaybeInputsOwned {
+        MaybeInputsOwned {
+            psbt: self.psbt,
+        }
+    }
+
+    /// Call this method if the only way to initiate a PayJoin with this receiver
+    /// requires manual intervention, as in most consumer wallets.
+    ///
+    /// So-called "interactive" receivers, like payment processors, that allow arbitrary requests are otherwise vulnerable to probing attacks.
+    /// Those receivers call `get_transaction_to_check_broadcast()` and `attest_tested_and_scheduled_broadcast()` after making those checks downstream.
+    pub fn attest_manual_receive_endpoint(self) -> MaybeInputsOwned {
+        MaybeInputsOwned {
+            psbt: self.psbt,
+        }
+    }
+}
+
+impl MaybeInputsOwned {
+    /// The receiver should not be able to sign for any of the Original PSBT's inputs.
+    /// Check that none of them are owned by the receiver downstream before proceeding.
+    pub fn input_script_pubkeys(&self) -> Vec<Result<&Script, RequestError>> {
+        todo!();
+    }
+
+    /// If the sender included inputs that the receiver could sign for in the original PSBT,
+    /// the receiver must either return error original-psbt-rejected or make sure they do not sign those inputs in the payjoin proposal.
+    ///
+    /// Call this after checking downstream.
+    pub fn attest_inputs_not_owned(self) -> MaybeScriptsSupported {
+        MaybeScriptsSupported {
             psbt: self.psbt,
         }
     }
+}
+
+impl MaybeScriptsSupported {
+    pub fn input_script_types(&self, network: bitcoin::Network) -> Vec<Option<AddressType>> {
```

Comment:
I don't think `network` will be needed here since there's nothing inside Bitcoin transaction that specifies the network. Also the return type should be an iterator.